### PR TITLE
Add bandit

### LINKED ISSUE #102: QGIS security scan blocks plugins using qgis_plugin_tools
QGIS security scan runs bandit checks for the plugins. Bandit reports 7 issues in this repo.

```
[main]  INFO    Found project level .bandit file: .\.bandit
[main]  INFO    Using command line arg for config file
[main]  INFO    Using command line arg for selected targets
[main]  INFO    profile include tests: None
[main]  INFO    profile exclude tests: None
[main]  INFO    cli include tests: None
[main]  INFO    cli exclude tests: None
[main]  INFO    using config: .\pyproject.toml
[main]  INFO    running on Python 3.12.14
Run started:2026-09-18 06:56:27.291266+00:00

Test results:
>> Issue: [B404:blacklist] Consider possible security implications associated with the subprocess module.
   Severity: Low   Confidence: High
   CWE: CWE-78 (https://cwe.mitre.org/data/definitions/78.html)
   More Info: https://bandit.readthedocs.io/en/1.9.4/blacklists/blacklist_imports.html#b404-import-subprocess
   Location: .\infrastructure\plugin_maker.py:8:0
7       import shutil
8       import subprocess
9       import sys

--------------------------------------------------
>> Issue: [B607:start_process_with_partial_path] Starting a process with a partial executable path
   Severity: Low   Confidence: High
   CWE: CWE-78 (https://cwe.mitre.org/data/definitions/78.html)
   More Info: https://bandit.readthedocs.io/en/1.9.4/plugins/b607_start_process_with_partial_path.html
   Location: .\infrastructure\plugin_maker.py:337:22
336             else:
337                 process = subprocess.Popen(
338                     "cmd.exe",
339                     shell=False,
340                     universal_newlines=True,
341                     stdin=subprocess.PIPE,
342                     stdout=sys.stdout,
343                     stderr=sys.stderr,
344                 )
345                 process.communicate(script)

--------------------------------------------------
>> Issue: [B603:subprocess_without_shell_equals_true] subprocess call - check for execution of untrusted input.
   Severity: Low   Confidence: High
   CWE: CWE-78 (https://cwe.mitre.org/data/definitions/78.html)
   More Info: https://bandit.readthedocs.io/en/1.9.4/plugins/b603_subprocess_without_shell_equals_true.html
   Location: .\infrastructure\plugin_maker.py:339:22
338                     "cmd.exe",
339                     shell=False,
340                     universal_newlines=True,
341                     stdin=subprocess.PIPE,
342                     stdout=sys.stdout,
343                     stderr=sys.stderr,
344                 )
345                 process.communicate(script)
346
347         def transup(self):

--------------------------------------------------
>> Issue: [B603:subprocess_without_shell_equals_true] subprocess call - check for execution of untrusted input.
   Severity: Low   Confidence: High
   CWE: CWE-78 (https://cwe.mitre.org/data/definitions/78.html)
   More Info: https://bandit.readthedocs.io/en/1.9.4/plugins/b603_subprocess_without_shell_equals_true.html
   Location: .\infrastructure\plugin_maker.py:407:18
406                 "cmd.exe" if is_windows() else "sh",
407                 shell=False,
408                 universal_newlines=True,
409                 stdin=subprocess.PIPE,
410                 stdout=sys.stdout,
411                 stderr=sys.stderr,
412                 env=env,
413             )
414             process.communicate(script)
415
416         @staticmethod

--------------------------------------------------
>> Issue: [B603:subprocess_without_shell_equals_true] subprocess call - check for execution of untrusted input.
   Severity: Low   Confidence: High
   CWE: CWE-78 (https://cwe.mitre.org/data/definitions/78.html)
   More Info: https://bandit.readthedocs.io/en/1.9.4/plugins/b603_subprocess_without_shell_equals_true.html
   Location: .\infrastructure\plugin_maker.py:422:15
421             echo(cmd, force=force_show_output)
422             pros = subprocess.Popen(
423                 args,
424                 cwd=d,
425                 stdout=subprocess.PIPE,
426                 stderr=subprocess.PIPE,
427                 universal_newlines=True,
428             )
429             stdout, stderr = pros.communicate()

--------------------------------------------------
>> Issue: [B110:try_except_pass] Try, Except, Pass detected.
   Severity: Low   Confidence: High
   CWE: CWE-703 (https://cwe.mitre.org/data/definitions/703.html)
   More Info: https://bandit.readthedocs.io/en/1.9.4/plugins/b110_try_except_pass.html
   Location: .\tools\custom_logging.py:194:8
193                     )
194             except Exception:
195                 pass
196

--------------------------------------------------
>> Issue: [B101:assert_used] Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.
   Severity: Low   Confidence: High
   CWE: CWE-703 (https://cwe.mitre.org/data/definitions/703.html)
   More Info: https://bandit.readthedocs.io/en/1.9.4/plugins/b101_assert_used.html
   Location: .\tools\resources.py:143:16
142                 if is_plugin := _is_module_qgis_plugin(module_name):
143                     assert is_plugin.plugin_directory
144                     return is_plugin.plugin_directory

--------------------------------------------------

Code scanned:
        Total lines of code: 2825
        Total lines skipped (#nosec): 0

Run metrics:
        Total issues (by severity):
                Undefined: 0
                Low: 7
                Medium: 0
                High: 0
        Total issues (by confidence):
                Undefined: 0
                Low: 0
                Medium: 0
                High: 7
Files skipped (0):
```

### PR DESCRIPTION
**Pull request by LKajan**
## Summary

Adds bandit check to pre-commit hooks and fixes the 7 bandit findings.

- `plugin_maker.py`: resolve the shell via `COMSPEC`/`shutil.which` instead of a bare `cmd.exe`, and mark the internal `subprocess` calls with `# nosec` with a justification comment.
- `custom_logging.py`: change too wide catch `except Exception: pass` with `contextlib.suppress(RuntimeError)`. **Behavioral change**
- `resources.py`: replace an `assert` used for type narrowing with a real check. **Small behavioral change** `resources.plugin_path()` no longer fails with an `AssertionError` when a plugin module has no directory. It now falls back to the default directory tree.

Assisted-by: Claude Code

## Related issues

Resolves #102

## How this was tested

Pre-commit run locally. Bandit don't report issues anymore.

Pytests run on QGIS3 environment passes. On QGIS4 environment 2 tests fails (I'll provide a separate issue and PR for this).

## Contributor checklist

- [x] I have read the [CONTRIBUTING guidelines](https://github.com/osgeosuomi/.github/blob/main/CONTRIBUTING.md).
- [x] **I have reviewed and understand every change in this pull request, and I take responsibility for it as the author.** This applies whether the changes were written by hand or with the assistance of AI tools.
- [x] If AI tools were used to generate a substantial part of this contribution, I have disclosed it in the PR description above or via an `Assisted-by:` commit trailer.
- [x] My contribution is licensed under the same license as this repository, and I have the right to submit it.
- [x] I have run the project's tests locally where applicable.

### CODE REVIEW

File: tools/resources.py
Comment (Joonalai) [inline tools/resources.py:130]: Should return value be `str | None` now?

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -22,6 +22,12 @@ repos:
       - id: mypy
         additional_dependencies:
           - types-requests
+  - repo: https://github.com/PyCQA/bandit
+    rev: 1.9.4
+    hooks:
+      - id: bandit
+        args: [ -c, pyproject.toml ]
+        additional_dependencies: [ "bandit[toml]" ]
```
Comment (Joonalai) [inline .pre-commit-config.yaml:30]: Does it really need `bandit[toml]` and `.bandit` file? In my other project, I [added bandit](https://github.com/Joonalai/profiler-qgis-plugin/commit/08b87cac1c8b5f49fa9837ff948352ebbd9479ec) without these and it still was able to read pyproject.toml (or did I miss something there?).

**Review by Joonalai — APPROVED** [review]
Few comments, otherwise looks good.

File: tools/resources.py
Comment (LKajan) [reply tools/resources.py:130]: Nothing changed regarding this.
The `is_plugin.plugin_directory` cannot be None at the return because the `if` and the `_plugin_path_submodule()` should always be str.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (LKajan) [reply .pre-commit-config.yaml:30]: It works without the `[toml]` on new Pythons but not if run on python before toml was added to standard library (py 3.11).

Bandit don't automatically catch the config from pyproject.toml. It must be given with the `-c`. When running manually, I'm lazy and have bad memory so I added the `.bandit` so `uv run bandit` works directly. For pre-commit the pyproject.toml must still be given as parameter because it doesn't see the .bandit somewhy.

Comment (Joonalai) [reply .pre-commit-config.yaml:30]: Alright, fair enough :smile:

**Comment by LKajan** [comment]
@Joonalai Do you mind if I fix the failing CI in this PR or do you prefer another PR for that? The QGIS image don't have git installed so it should be a small fix.

**Comment by Joonalai** [comment]
> @Joonalai Do you mind if I fix the failing CI in this PR or do you prefer another PR for that? The QGIS image don't have git installed so it should be a small fix.

Not at all, it would be great if you could fix that here.

**Comment by Joonalai** [comment]
And one more thing: add something to CHANGELOG.md, I can then make a new release afterwards.

**Comment by LKajan** [comment]
> And one more thing: add something to CHANGELOG.md, I can then make a new release afterwards.

Done.
I noticed the change in the `resources.plugin_path()` should be counted as a behavioral change as well. If a plugin module has no directory it now falls back to the default directory and don't fail to AssertionError anymore. Should that actually raise a custom exception instead?

**Comment by Joonalai** [comment]
> I noticed the change in the `resources.plugin_path()` should be counted as a behavioral change as well. If a plugin module has no directory it now falls back to the default directory and don't fail to AssertionError anymore. Should that actually raise a custom exception instead?

Yes, I was also thinking that it might break something if no error is thrown. Raise a custom exception so nothing will be broken.

**Comment by LKajan** [comment]
Actually the old assert was just a hint for type checkers, couldn't be raised in reality. In that code path when the `is_plugin` is True the `IsPluginResult` contains alwas the directory. So, no change in behavior after all.
